File: 0261-graph-valid-tree/0261-graph-valid-tree.cpp
// TC- O(v+e)
// SC- O(e+v)

class Solution {
private:
    bool hascycle(vector<vector<int>> &adj, unordered_set<int>& visited, int n, int parent){
        
        visited.insert(n);
        for(int neigh : adj[n]) {
            if(neigh == parent)
                continue;

            if(visited.find(neigh) != visited.end())
                return true;

            if(hascycle(adj, visited, neigh, n))
                return true;
    }
        return false;
    }
public:
    bool validTree(int n, vector<vector<int>>& edges) {
        //create adj list
        vector<vector<int>> adj(n);
        for(auto edge:edges){
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
        }

        // create list for visit state for detecting cycle and int for components
        unordered_set<int> visited;
        int component = 0;

        // dfs
        for(int i=0; i<n; i++){
            if(visited.find(i)==visited.end()){
                component++;
                if(component>1)
                    return false;
                if(hascycle(adj, visited, i, -1))
                    return false;
            }
        }
        return true;
    }
};

File: 0261-graph-valid-tree/0261-graph-valid-tree.py
# TC- O(V+E)
# SC- O(V+E)

class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        # valid tree if graph with no cycles, n-1 edges and is fully connected.

        if len(edges)!=n-1:
            return False

        # create adj mat
        adj_mat = defaultdict(list)
        for edge in edges:
            adj_mat[edge[0]].append(edge[1])
            adj_mat[edge[1]].append(edge[0])

        visited = set()
        def has_cycle(node, parent):
            visited.add(node)
            for nei in adj_mat[node]:
                if nei==parent:
                    continue
                elif nei in visited:
                    return True
                elif has_cycle(nei, node):
                    return True
            return False

        if has_cycle(0, -1):
            return False

        return len(visited) == n
